plugins/BEdita/Core/src/Job/ServiceRunner.php

```diff
@@ -0,0 +1,166 @@
+        }
+        $instance = new $classFound();
+        static::register($name, $instance);
+
+        return $instance;
+    }
+
+    /**
+     * Register service class instance for a $name service.
+     * Instance MUST implement JobService
+     *
+     * @param string $name The service name you want to register.
+     * @param mixed $instance The instance object to be registered, MUST implement JobService
+     * @return void
+     * @throws \LogicException
+     */
+    public static function register($name, $instance)
+    {
+        if (!($instance instanceof JobService)) {
+            Log::write('error', 'bad service class: ' . get_class($instance));
+            throw new \LogicException(__d('bedita', 'Bad service class "{0}"', [get_class($instance)]));
+        }
+        static::$instances[$name] = $instance;
+    }
+
+    /**
+     * Reset registered service instances
+     *
+     * @return void
+     */
+    public static function reset()
+    {
+        static::$instances = [];
+    }
+
+    /**
+     * Run a service with $payload data read from async jobs table
+     *
+     * In $options array this parameters are available:
+     *  - 'lockPeriod' : lock time duration, specify in the form '+5 minutes'
+     *
+     * @param string $uuid UUID of the job to run.
+     * @param array $options Options for running this job.
+     * @return bool True on success, false on failure
+     */
+    public static function run($uuid, $options = [])
+    {
+        $locked = false;
+        $AsyncJobs = TableRegistry::get('AsyncJobs');
+        try {
+            $job = $AsyncJobs->lock($uuid, Hash::get($options, 'lockPeriod', '+5 minutes'));
+            $locked = true;
+            $service = static::getService($job->service);
+            $success = $service->run($job->payload, $options);
+            $AsyncJobs->unlock($uuid, $success);
+
+            return $success;
+        } catch (\Exception $e) {
+            Log::write('error', 'job run failed: ' . $e->getMessage());
+            if ($locked) {
+                // locked job failed, unlock
+                $AsyncJobs->unlock($uuid, false);
+            }
+
+            return false;
+        }
+    }
+
+    /**
+     * Run pending jobs reading from async jobs table.
+     * Using an optional max number of jobs as 'limit'
+     * Resulting array will contain
+     *  - 'count' number oj jobs executed
+     *  - 'success' uuids of jobs successfully executed
+     *  - 'success' uuids of failed jobs
+     *
+     * @param int $limit Max number of pending jobs to run.
+     * @return array Result details
+     */
+    public static function runPending($limit = 0)
+    {
+        $results = ['success' => [], 'failure' => []];
+        $pending = TableRegistry::get('AsyncJobs')->find('pending')->select(['uuid']);
+        if ($limit) {
+            $pending->limit($limit);
+        }
+        $count = 0;
+        foreach ($pending as $job) {
```

Comment:
(To be investigated in the future) PDO allows query results to be streamed instead of being buffered (see [working with result sets](https://book.cakephp.org/3.0/en/orm/retrieving-data-and-resultsets.html#working-with-result-sets)). This could help when running pending jobs in "unlimited" mode and a very large queue.